osbenchmark/worker_coordinator/runner.py

```diff
@@ -967,6 +968,75 @@ async def _scroll_query(opensearch, params):
                 "took": took
             }
 
+        async def _vector_search_query_with_recall(opensearch, params):
+            """
+            Perform vector search and report recall@k , recall@r and time taken to perform recall in ms as
+            meta object.
+            """
+            result = {
+                "weight": 1,
+                "unit": "ops",
+                "success": True,
+            }
+
+            def recall_at_r(results, neighbors, r):
+                """
+                Calculates the recall@R for a set of queries against a ground truth nearest
+                neighbor set
+                Args:
+                    results: list containing ids of results returned by OpenSearch.
+                    neighbors: set containing ids of the actual neighbors
+                    neighbors for a set of queries
+                    r: number of top results to check from the neighbors
+                Returns:
+                    Recall at R
+                """
+                correct = 0.0
+                if neighbors is None:
+                    return 0.0
+                total_num_of_results = min(r, len(neighbors))
+                for j in range(total_num_of_results):
+                    if results[j] in neighbors:
+                        correct += 1.0
+
+                return correct / total_num_of_results
+
+            doc_type = params.get("type")
+            response = await self._raw_search(opensearch, doc_type, index, body, request_params, headers=headers)
+            recall_processing_start = time.perf_counter()
+            response_json = json.loads(response.getvalue())
+            if detailed_results:
+                props = parse(response, ["hits.total", "hits.total.value", "hits.total.relation", "timed_out", "took"])
+                hits_total = props.get("hits.total.value", props.get("hits.total", 0))
+                hits_relation = props.get("hits.total.relation", "eq")
+                timed_out = props.get("timed_out", False)
+                took = props.get("took", 0)
+
+                result.update({
+                    "hits": hits_total,
+                    "hits_relation": hits_relation,
+                    "timed_out": timed_out,
+                    "took": took
+                })
```

Comment:
what will happen here if hits are empty? and search is giving 200.

Reply:
Updated.

Reply:
Added test case where hits contain empty list.